### ISSUE TITLE
GODOT 3.0 BETA APPROACHING!

### ISSUE THREAD

**Issue by reduz** [godotengine/godot#12543]
Hi guys! Now that Alpha2 is out, I'm using a GitHub issue to communicate our current plans for beta.

The tentative date is November 20th. This is the **last chance** to get work, features, improvements, etc. merged. After this, we go full freeze and **only bug fixes** will be merged.

If you are working on something that needs to be completed, feel free to join #godotengine-beta on freenode IRC, so we can organize this work.

**Comment by reduz** [comment]
It is advised you use an irc client that allows a persistent connection so you don't miss information on this channel. Try using matrix (http://matrix.org) or Irccloud (http://irccloud.com)

**Comment by NullNeska** [comment]
Does converting text resources to binary on exporting count as a feature or a bug fix, in this case.

**Comment by reduz** [comment]
that is a fix

**Comment by Zireael07** [comment]
Hoping for Bullet physics toggle to make the cut *hint hint*

**Comment by NathanWarden** [comment]
@Zireael07 I second that! :)

**Comment by PaulTaulborg** [comment]
- We need 3.0 and its physics to work with all monitor refresh rates.

- We need websocket support added to the core ASAP.

- We need BitMap font support for 3.0

- We need a centralized location instead of using github as a discussion forum.  Facebook is not okay and  terrible for Q&A style posts.

- We need easier access for people to send in documentation changes. People should not have to be forced to  fork an entire repo to add a  sentence for a method.

- We need ability for users to add a custom icon for their executable, not have to rely on resource hacker

- We need more UI .cpp classes that are centered around game engine stuff. Tetris style inventory system, skill tree system, other inherent game UI features.  **Godot really shines in the 2d department, we need to make it shine brighter.**

**Comment by reduz** [comment]
> We need 3.0 and its physics to work with all monitor refresh rates.

Not gonna happen for 3.0

> We need 3.0 and its physics to work with all monitor refresh rates.

This is being worked on, AFAIK, may or may not make it for 3.0

> We need a centralized location instead of using github as a discussion forum. Facebook is not okay and terrible for Q&A style posts.

Not gonna happen, ever.

> We need BitMap font support for 3.0

There is bitmap font support, just not as resources. Godot can load BMFont files.

> We need easier access for people to send in documentation changes. People should not have to be forced to fork an entire repo to add a sentence for a method.

Current way is OK

> We need ability for users to add a custom icon for their executable, not have to rely on resource hacker

Problem with this is that it seems to be only doable when exporting from Windows. Cant be done from other platforms. All libraries we tried to do this ended up corrupting the executable.

> We need more UI .cpp classes that are centered around game engine stuff. Tetris style inventory system, skill tree system, other inherent game UI features. Godot really shines in the 2d and UI department, we need to make it shine brighter.

Vote for it on Patreon, or encourage others to submit stuff like this to Asset Library. No game-specific code will be added to Godot.

**Comment by akien-mga** [comment]
Note: This is a thread of engine developers to discuss what are the remaining tasks to do for Godot 3.0 beta. This is **not** a thread for all community members to dump their wishlist, we're not Santa Claus.

**Comment by PaulTaulborg** [comment]
> No game-specific code will be added to Godot.

And what about the Sprite/AnimatedSprite nodes?  ![](http://forums.d2jsp.org/html/emoticons/wallbash.gif)

 
> Not gonna happen, ever.

Well, all questions and posts are not publicly indexed and available to future godot developers when googling. So all the help and support received in facebook will be hidden. That's a detriment and not fair to future Godot developers.


> There is bitmap font support, just not as resources. Godot can load BMFont files.

I mean the ability to use the actual font.  See here #10184 

> Problem with this is that it seems to be only doable when exporting from Windows. Cant be done from other platforms. All libraries we tried to do this ended up corrupting the executable.

That is fair. More of a windows fault, I can understand that.

> This is being worked on, AFAIK, may or may not make it for 3.0

Understood. I was just curious about that grant, that awarded you guys $20,000
>  "On 22 June 2016, Godot received a $20,000 Mozilla Open Source Support (MOSS) “Mission Partners” award to be used to add Web Sockets, WebAssembly and WebGL 2.0 support."

I expect the developer who is working on the Web Socket module to be partially paid then?


> Note: This is a thread of engine developers to discuss what are the remaining tasks to do for Godot 3.0 beta. This is not a thread for all community members to dump their wishlist, we're not Santa Claus.

And that is _not what the OP said at all._

**Comment by reduz** [comment]
> And what about the Sprite/AnimatedSprite nodes? 

That is a far lower level feature, and way more commonly used feature, compared to a ring-menu.

> I mean the ability to use the actual font. See here #10184

Probably not too difficult to do a BMFont inmporter now that we renamed the font extension, but there's so many other things to do. If you open an issue about specifically creating a BMFont importer, we may tag it junior job so anyone can give it a try

> Understood. I was just curious about that grant, that awarded you guys $20,000

All that work was for the HTML5 port (which consisted in a ton more work than that), and technically the TCP connections are routed over websockets internally in WebAssembly (vua the sockets to websockets conversion) so this could be considered done. An external websocket library to use independently was never promised, but there are users working on this anyway, so not exactly sure why are you so upset. 

I suggest you donate a good chunk of money to Patreon and vote for the features you want, or start doing some coding, because honestly I think you are being incredibly selfish here.

**Comment by PaulTaulborg** [comment]
> All that work was for the HTML5 port

Aww. I see it now here

https://blog.mozilla.org/blog/2016/06/22/mozilla-awards-385000-to-open-source-projects-as-part-of-moss-mission-partners-program/

It actually says: 
> building of compelling games on the Web.

Which is far different than the description on Wikipedia. It's missing that part. That makes sense now, my apologizes.

---------------
Here's the deal Juan. Listen closely please before I get myself banned by akien

I know you are going to rebute any suggestion I make. The hierarchy here is quite evident: 
"Godot Dev > regular person offering suggestions".

However, I am trying to tell you there is a core issue in this engine that needs to be prioritized! All the suggestions and ideas I give, sure, disagree with me all you want. **And I respect your decision.**  

All I am trying to say, anyone who uses physics and has a 61hz or higher monitor will have jitter, stutter, smooth movement issues.  I am baffled on why this is not prioritized at all. And instead, C# gets added, more 3d work, etc, etc.

**I want the best for Godot** and just feel like the Godot _devs as a whole_ are not listening to me about this. 

**This is a huge issue**. because Godot developers who spend time creating their game using the current physics and the Camera2D node now, and then ship their game, guess what? Any player who has a 61hz or higher monitor = will have jitter.  

The solution could be.. well just update your character as a kinematic body under _process. However, camera2d runs at 60fps, so.. it would be out of sync and create jitter. What current developers need to do is: create a custom camera_transform, update that in `_process()`, and use a  `kinematicbody`. That makes it works on all refresh monitor rates. 

However, that's a kinematic body, not a rigid one. We cannot use rigidbody physics because all that is tied under `fixed_fps` and there is no way to get that with the same rate as the monitor. In fact, I remember reading here that changing the `fixed_fps` is a bad idea.

And now, since 3.0 is approaching.. you said:

> Not gonna happen for 3.0

Which is just crazy to me. I think you are deeply underestimating how many players out there have higher than 61hz monitors. I am pleading you to try to prioritize this. I am not trying to be selfish and apologize for coming off like that. I am a patreon btw (not that it should matter since my donation amount is low anyway), but still.

**Comment by reduz** [comment]
> Which is just crazy to me. I think you are deeply underestimating how many players out there have higher than 61hz monitors.

And I think you are deeply understimating how expensive 144hz  monitors are where I live. I may get one from Europe in February from Godotcon, so I can better evaluate different solutions to the problem.

**Comment by PaulTaulborg** [comment]
Yes, Juan check this out, I remember seeing this in discord: https://i.gyazo.com/b8dbdf2628b0a53cb83067d8ec0c28b9.png
You can [jump](https://discordapp.com/channels/212250894228652034/212250894228652034?jump=370822154520100865) to the conversation if you want to read more. He basically saying you can still test it, if you have a 60hz monitor.  

And I never knew about those prices in Argentina for monitors, that's ridiculous! It kind of now makes sense why you have been hesitant with it.

**Comment by reduz** [comment]
@PaulTaulborg As mentioned, there may be other ways to mitigate the problem in monitors with high refresh, so I prefer to see the problem myself and decide.

**Comment by Zylann** [comment]
Some things in the terrains tracker? https://github.com/godotengine/godot/issues/9046

**Comment by Alexia-AT-Digitecnology** [comment]
@AndreaCatania needs to merge the PR of bullet integration before Godot 3.0 beta.

**Comment by AndreaCatania** [comment]
@PaulTaulborg Hi, can you please create a issue that explain this problem and upload a sample project where I can test it?

Don't forgot to tag me and Reduz. Thanks

**Comment by kubecz3k** [comment]
@AndreaCatania I dont think the answer for the problem is inside physics engine. We should implement fixed step interpolation instead (jitter is very noticeable on any framerate different than screen refresh rate, and that means if someone has slower pc, and the game is able to run only with for example 55fps, it will jitter).

**Comment by AndreaCatania** [comment]
@kubecz3k yes I know. Thanks for #10388

**Comment by sarahspock** [comment]
Recommend SteamVR or OpenVR features for 3.0.

**Comment by OvermindDL1** [comment]
Steam/OpenVR is actually pretty easy to implement.  Need to add in a special full screen shader (render the top framebuffer with it) to 'bend' the scene properly, and then just handle the inputs as necessary.  Perhaps not for 3.0 but it would be *very* nice to add (and fairly easy) and would gain interest in Godot from the VR world.

**Comment by Zireael07** [comment]
Folks, this is not a place to request 'insert feature here', especially if no PR has been made for it.

**Comment by BastiaanOlij** [comment]
Guys, OpenVR is already done: https://github.com/BastiaanOlij/godot_openvr
If you build the alpha2 branch you have a copy that works with Alpha 2

It's a GDNative module on top of the ARVR framework in Godot 3.

See also: https://www.youtube.com/watch?v=v291rMWCMRw&t=39s

Finally I'll be updating builds for Alpha 2 tonight so you can download the module without having to compile it (we found out yesterday that the one I had online only works if you have Visual Studio installed due to dependencies with the debug build).

**Comment by OvermindDL1** [comment]
@BastiaanOlij  Oh now that is awesome, I'll have to give that a try (though on linux).  ^.^

**Comment by BastiaanOlij** [comment]
I haven't got a linux setup but Ben's been testing it on Linux I believe. You will need to compile openvr yourself.

**Comment by RodeoMcCabe** [comment]
@PaulTaulborg 

> We need a centralized location instead of using github as a discussion forum. Facebook is not okay and terrible for Q&A style posts.

Centralized location for a discussion forum, you say? [https://godotdevelopers.org/forum/](https://godotdevelopers.org/forum/) How about that.

Q&A style posts, you say? [https://godotengine.org/qa/](https://godotengine.org/qa/) How about that.

**Comment by satyries** [comment]
What about LOD?
When dealing with 3D this is a very _traversal_  feature: it come handy to smartphone app to full desktop games.

**Comment by akien-mga** [comment]
Guys, next comment requesting another feature, I lock this issue for non-contributors. Quoting myself again:

> Note: This is a thread for engine developers to discuss what are the remaining tasks to do for Godot 3.0 beta. This is not a thread for all community members to dump their wishlist, we're not Santa Claus. Please respect that, otherwise we'll have to lock discussions to core developers to keep things manageable. If there are issues that you think are going unnoticed, feel free to comment in those issues directly.

Only people involved in actually developing the features which are planned (and almost completely implemented already) for Godot 3.0 should be talking here. We're not asking for ideas of new features to cram in at the last minute... Or do you really want Godot 3.0 to release in 2019 so that we can implement all your suggestions?

**Comment by akien-mga** [comment]
There will be other releases after 3.0, don't try to force all your favorite features in an already way too big release that has been in the making for 18 months.

**Comment by RandomShaper** [comment]
I would be able to finish basic onion skinning (#12572) if the engine provided a way of copying textures from VRAM to VRAM.

The only remaining task would be to add something to the UI to enable/disable it.

Even auto-keyframing (or another way of previewing how modifications in the present affect past frames) may be implemented in time. But basic support adds some value by itself.

**Comment by Ranoller** [comment]
> 

Which is just crazy to me. I think you are deeply underestimating how many players out there have higher than 61hz monitors.And I think you are deeply understimating how expensive 144hz monitors are where I live. I may get one from Europe in February from Godotcon, so I can better evaluate different solutions to the problem.
--

This is the second time that i read you complaining about money. We (patreon suporters) put money to this kind of things not happend. If you need aditional money to buy a 144hz monitor for you, akien,  other devs, buy one mac, buy one windows system, one android per dev, one iphone per dev, etc.... PLEASE DO A KIKSTARTER or something similar. Is hoppeles that one problem, or one feature, will not be tested because not getting simple "stuff", I believe that the godot community is responding well enough to income requests, as to respond that things are not done for lack of income.

Please do not use lack of money excuses without asking for this money, because this is a dead end, and is not good to godot public image.

**Comment by reduz** [comment]
@Ranoller If it helps, I stil haven't received a dime from the donations, and it will take a while to sort out because we are doing properly and legally. 

Also, how I use the money is not of your concern. If I want to save money by bringing one from Europe is my problem, not yours. Donating does not give you the right to extort me to implement whatever feature you want. We have a system in place for voting this and, if you don't like it, no one is forcing you to pledge.

**Comment by reduz** [comment]
@Ranoller Also, let me remind you that, with Conservancy, we have a directed donation system in place. You can donate to Conservancy directly and that money will be used to hire a programmer to work on whatever feature you need (as long as it's agreed that the project needs it). 

This is how the Mozilla and Microsoft donations took place. Of course, hiring a programmer costs a lot more than just pledging to Patreon. If you want to organize a Kickstarter yourself to fund this, feel welcome. Otherwise, just be patient a wait a while until this gets solved.

**Comment by Ranoller** [comment]
@reduz 
Of course your money is yours. I´m not extorting you to any feature, i´m talking about "the monitor". You are missunderstanding my mensagge, but i think that deeply you now what i mean. It´s clear that donating is donating: no aditional right´s to "no donating" user except vote and another rewards, I´m not the user that ask the "144hz" physiscs problem solution, igualmente si después de la godotcon te pasas por España me acerco contigo a por el monitor y te invito a unas cañas.

**Comment by reduz** [comment]
@Ranoller In fact, my main monitor died a week ago and I seriously considered getting a 144hz one, but it's really expensive (around 1000usd vs 150usd for a regular one) so I ended up buying a crappier one. I know they are much cheaper in Europe and given I'll be travelling for the godotcon I prefer to get one there. Tal vez no lo parezca, pero la vida en Buenos Aires es mucho mas costosa que en España (donde estuve hace unos meses). Todo desde la comida hasta lo electrónico sale 3 veces mas, no es que sobra el dinero por aquí.

**Comment by Ranoller** [comment]
@reduz No se en francia (¿es allí la godotcon?) cuanto cuestan. Aquí por amazon lo puedes comprar por 280 euros más o menos que serían unos 320USD. Entiendo que ahora godot pasa por un impás de "indie" a "mainstream", ya verás cuando todo se consolide como esos mil dolares te parecen una nadería. ;)

**Comment by neikeq** [comment]
Guys, stop putting Akien down in spanish.

**Comment by JunShiozawa** [comment]
Which Does Godot 3.0 need minSdkVersion on Android?

**Comment by RandomShaper** [comment]
Onion skinning is moving forward. I'd need some advice. Please have a look at its thread.

**Comment by mrezai** [comment]
@reduz @akien-mga 
Please consider this PR for beta: #10546

**Comment by volzhs** [comment]
@reduz  what do you think about #12256 before beta? `post_import` is really helpful.

**Comment by akien-mga** [comment]
BTW to all designers (@djrm @toger5 @volzhs and co): if you still want to make theme and usability changes, it should be done before beta. Afterwards we need the interface to stop changing so that documentation writers can take screenshots.

**Comment by djrm** [comment]
For me the interface is 'ready', it has some quirks but if those are changed they wont affect documentation material, on the visual side i think its ready (no changes to VS, or major icon or theme changes).

**Comment by toger5** [comment]
@akien-mga  agree with djrm

**Comment by djrm** [comment]
@akien-mga actually there is something, i proposed changing the TileMap and GridMap palette to the right, not a big change, but it would be nice if it was considered for the beta.  #12488

**Comment by akien-mga** [comment]
> actually there is something, i proposed changing the TileMap and GridMap palette to the right, not a big change, but it would be nice if it was considered for the beta. #12488

Best way to decide is to merge and see if people complain, so let's go :D

**Comment by djrm** [comment]
the same has to be done for TileMap

**Comment by RandomShaper** [comment]
The date for the feature freeze is almost here.

Once more about onion skinning, #12572 is ready for testing and even merging in its current state. I'd like to improve some aspects of it, but I could do it before the final release as well as any bug fix needed, if it got merged now.

**Comment by amirea** [comment]
> No game-specific code will be added to Godot
Ojalá! (Sorry, I couldn't help). This principle makes Godot streamlined and preserves users' freedom. I'm confident Godot will keep resisting contamination (as opposed to, i.e. Unreal Engine).

**Comment by akien-mga** [comment]
Fixed :) https://godotengine.org/article/dev-snapshot-godot-3-0-beta-1
